src/pkg/systemartifact/manager.go

```diff
@@ -0,0 +1,253 @@
+			return err
+		}
+		repoName := mgr.getRepositoryName(artifactRecord.Vendor, artifactRecord.Repository)
+		err = mgr.regCli.PushBlob(repoName, artifactRecord.Digest, artifactRecord.Size, reader)
+		if err != nil {
+			return err
+		}
+		artifactId = id
+		return nil
+	})(ctx)
+
+	return artifactId, createError
+}
+
+func (mgr *systemArtifactManager) Read(ctx context.Context, vendor string, repository string, digest string) (io.ReadCloser, error) {
+	sa, err := mgr.dao.Get(ctx, vendor, repository, digest)
+	if err != nil {
+		return nil, err
+	}
+	repoName := mgr.getRepositoryName(vendor, repository)
+	_, readCloser, err := mgr.regCli.PullBlob(repoName, sa.Digest)
+	if err != nil {
+		return nil, err
+	}
+	return readCloser, nil
+}
+
+func (mgr *systemArtifactManager) Delete(ctx context.Context, vendor string, repository string, digest string) error {
+
+	repoName := mgr.getRepositoryName(vendor, repository)
+	if err := mgr.regCli.DeleteBlob(repoName, digest); err != nil {
+		log.Errorf("Error deleting system artifact BLOB : %s. Error: %v", repoName, err)
+		return err
+	}
+
+	return mgr.dao.Delete(ctx, vendor, repository, digest)
+}
+
+func (mgr *systemArtifactManager) Exists(ctx context.Context, vendor string, repository string, digest string) (bool, error) {
+	_, err := mgr.dao.Get(ctx, vendor, repository, digest)
+	if err != nil {
+		return false, err
+	}
+
+	repoName := mgr.getRepositoryName(vendor, repository)
+	exist, err := mgr.regCli.BlobExist(repoName, digest)
+
+	if err != nil {
+		return false, err
+	}
+
+	return exist, nil
+}
+
+func (mgr *systemArtifactManager) GetStorageSize(ctx context.Context) (int64, error) {
+	return mgr.dao.Size(ctx)
+}
+
+func (mgr *systemArtifactManager) RegisterCleanupCriteria(vendor string, artifactType string, criteria Selector) {
+	key := fmt.Sprintf(keyFormat, vendor, artifactType)
+	defer mgr.lock.Unlock()
+	mgr.lock.Lock()
+	mgr.cleanupCriteria[key] = criteria
+}
+
+func (mgr *systemArtifactManager) GetCleanupCriteria(vendor string, artifactType string) Selector {
+	key := fmt.Sprintf(keyFormat, vendor, artifactType)
+	defer mgr.lock.Unlock()
+	mgr.lock.Lock()
+	if criteria, ok := mgr.cleanupCriteria[key]; ok {
+		return criteria
+	}
+	return DefaultSelector
+}
+
+func (mgr *systemArtifactManager) Cleanup(ctx context.Context) (int64, int64, error) {
+	logger.Info("Starting system artifact cleanup")
+	// clean up artifact records having customized cleanup criteria first
+	totalReclaimedSize := int64(0)
+	totalRecordsDeleted := int64(0)
+
+	// get a copy of the registered cleanup criteria and
+	// iterate through this copy to invoke the cleanup
+	registeredCriteria := make(map[string]Selector, 0)
+	mgr.lock.Lock()
+	for key, val := range mgr.cleanupCriteria {
+		registeredCriteria[key] = val
+	}
+	mgr.lock.Unlock()
+
+	for key, val := range registeredCriteria {
```

Comment:
This will be better to understand?

```
for key, val := range registeredCriteria {
		logger.Infof("Executing cleanup for 'vendor:artifactType' : %s", key)
		deleted, size, err := mgr.cleanup(ctx, val)

		if err != nil {
			// one vendor error should not impact the clean-up of other vendor types. Hence the cleanup logic would continue
			// after logging the error
			logger.Errorf("Error when cleaning up system artifacts for 'vendor:artifactType':%s, %v", key, err)
                        continue
		}

		totalRecordsDeleted += deleted
		totalReclaimedSize += size
}
```